File: solutions/python/linked-list/1/linked_list.py
class Node:
    def __init__(self, value: int, succeeding: 'Node' = None, previous: 'Node' = None):
        self.value = value
        self.succeeding = succeeding
        self.previous = previous

    def __repr__(self):
        return f"{self.value}"


class LinkedList:
    def __init__(self):
        self.head_node = Node(0)
        self.tail_node = Node(0)
        self.head_node.succeeding = self.tail_node
        self.tail_node.previous = self.head_node
        self.count = 0

    def __len__(self):
        return self.count

    def push(self, value: int) -> None:
        previous = self.tail_node.previous
        node = Node(value, previous=previous, succeeding=self.tail_node)
        previous.succeeding = node
        self.tail_node.previous = node
        self.count += 1

    def unshift(self, value: int) -> None:
        succeeding = self.head_node.succeeding
        node = Node(value, previous=self.head_node, succeeding=succeeding)
        succeeding.previous = node
        self.head_node.succeeding = node
        self.count += 1

    def pop(self) -> int:
        if self.count == 0:
            raise IndexError("List is empty")
        node = self.tail_node.previous
        self.remove(node)
        return node.value

    def shift(self) -> int:
        if self.count == 0:
            raise IndexError("List is empty")
        node = self.head_node.succeeding
        self.remove(node)
        return node.value

    def delete(self, value) -> None:
        if not (node := self.find(value)):
            raise ValueError("Value not found")
        self.remove(node)

    def remove(self, node) -> None:
        previous = node.previous
        succeeding = node.succeeding
        previous.succeeding = succeeding
        succeeding.previous = previous
        self.count -= 1

    def find(self, value) -> Node:
        node = self.head_node.succeeding
        while node != self.tail_node:
            if node.value == value:
                return node
            node = node.succeeding
        return None
            
            

File: solutions/typescript/linked-list/1/linked-list.ts
type NodeRef<TElement> = Node<TElement> | undefined;

class Node<TElement> {

  value: TElement;
  next: NodeRef<TElement>;
  prev: NodeRef<TElement>;

  constructor (value: TElement) {
    this.value = value;
    this.next = undefined;
    this.prev = undefined;
  }
}

export class LinkedList<TElement> {

  head: NodeRef<TElement> = undefined;
  tail: NodeRef<TElement> = undefined;
  
  public push(element: TElement) {
    let node = new Node(element);
    let tail = this.tail;
    if (tail) {
      tail.next = node;
      node.prev = tail;
    }
    else {
      this.head = node;
    }
    this.tail = node;
  }

  public pop(): TElement {
    let tail = this.tail;
    if (tail) {
      this.#remove(tail);
      return tail.value;
    }
    throw new Error('List is empty');
  }

  public shift(): TElement {
    let head = this.head;
    if (head) {
      this.#remove(head);
      return head.value;
    }
    throw new Error('List is empty');
  }

  public unshift(element: TElement) {
    let node = new Node(element);
    let head = this.head;
    if (head) {
      head.prev = node;
      node.next = head;
    }
    else {
      this.tail = node;
    }
    this.head = node;
  }

  public delete(element: TElement) {
    let node = this.#find(element);
    if (node)
      this.#remove(node);
  }

  #find(element: TElement): NodeRef<TElement> {
    let node = this.head;
    while (node) {
      if (node.value == element)
        return node;
      node = node.next;
    }
    return undefined;
  }

  #remove(node: Node<TElement>) {
    let next = node.next;
    let prev = node.prev;
    if (prev)
      prev.next = next;
    if (next)
      next.prev = prev;
    if (this.head == node)
      this.head = next;
    if (this.tail == node)
      this.tail = prev;
  }

  public count(): number {
    let result = 0;
    let node = this.head;
    while (node) {
      result++;
      node = node.next;
    }
    return result;
  }
}
